packages/angular/ssr/src/routes/route-config.ts

```diff
@@ -0,0 +1,146 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.dev/license
+ */
+
+import { EnvironmentProviders, InjectionToken, makeEnvironmentProviders } from '@angular/core';
+
+/**
+ * Different rendering modes for server routes.
+ * @developerPreview
+ */
+export enum RenderMode {
+  /** AppShell rendering mode, typically used for pre-rendered shells of the application. */
+  AppShell,
+
+  /** Server-Side Rendering (SSR) mode, where content is rendered on the server for each request. */
+  SSR,
+
+  /** Client-Side Rendering (CSR) mode, where content is rendered on the client side in the browser. */
+  CSR,
+
+  /** Static Site Generation (SSG) mode, where content is pre-rendered at build time and served as static files. */
+  SSG,
+}
+
+/**
+ * Fallback strategies for Static Site Generation (SSG) routes.
+ * @developerPreview
+ */
+export enum SSGFallback {
+  /** Use Server-Side Rendering (SSR) as the fallback for this route. */
+  SSR,
+
+  /** Use Client-Side Rendering (CSR) as the fallback for this route. */
+  CSR,
+
+  /** No fallback; Angular will not handle the response if the path is not pre-rendered. */
+  None,
+}
+
+/**
+ * Common interface for server routes, providing shared properties.
+ */
+export interface ServerRouteCommon {
+  /** The path associated with this route. */
+  path: string;
+
+  /** Optional additional headers to include in the response for this route. */
+  headers?: Record<string, string>;
+
+  /** Optional status code to return for this route. */
+  status?: number;
+}
+
+/**
+ * A server route that uses AppShell rendering mode.
+ */
+export interface ServerRouteAppShell extends Omit<ServerRouteCommon, 'headers' | 'status'> {
+  /** Specifies that the route uses AppShell rendering mode. */
+  renderMode: RenderMode.AppShell;
+}
+
+/**
+ * A server route that uses Client-Side Rendering (CSR) mode.
+ */
+export interface ServerRouteCSR extends ServerRouteCommon {
+  /** Specifies that the route uses Client-Side Rendering (CSR) mode. */
+  renderMode: RenderMode.CSR;
+}
+
+/**
+ * A server route that uses Static Site Generation (SSG) mode.
+ */
+export interface ServerRouteSSG extends Omit<ServerRouteCommon, 'status'> {
+  /** Specifies that the route uses Static Site Generation (SSG) mode. */
+  renderMode: RenderMode.SSG;
+
+  /**
+   * Optional fallback strategy to use if the SSG path is not pre-rendered.
+   * Defaults to `SSGFallback.SSR` if not provided.
```

Comment:
Discussion: This is handling the case where a request matches an SSG route with a parameter not returned by `getPrerenderPaths` right?

I think we've discussed before what the right default fallback behavior is but I don't know that we came to any particular consensus. I want to at least consider whether `SSGFallback.None` would be a better default. My reasoning is that I expect most users don't think through this particular edge case or ask themselves whether or not it is ok to allow users to request arbitrary data in an SSG route at runtime. They might not expect that route to ever render on the server.

There are two potential counter points against `SSGFallback.None` I can think of:

First, the devserver ignores `RenderMode.SSG`, does not call `getPrerenderPaths`, and instead renders those routes like `SSR` on demand right? If so, then `SSGFallback.None` actually diverges the runtime behavior between the devserver (route is SSR'd because devserver can't know it's not provided by `getPrerenderPaths`) and production server (doesn't match, likely falls back to 404).

Second, this same case also exists with CSR. If the user visits any page and then performs a client-side navigation to a missing SSG route, the app router doesn't know the route is invalid and will attempt to render it. This diverges the initial load behavior for `SSGFallback.None` routes which don't match (likely 404) on the server but would attempt to render on the client. Client-side and server-side routing are already a bit diverged anyways, so this isn't the end of the world, but I do suspect it is a good idea to at least make sure they agree on what route ultimately gets rendered.

While this problem does technically exist with CSR, it's pretty difficult for users to encounter unless they find a link which actually redirects them to an SSG route with an invalid param. I think you'd need to inject some code in the browser console to get it to do a client-side navigation to such a route without an existing link you can click on. For SSR, this is a bit easier to walk into, given that users can type an invalid SSG param in the browser URL directly. Maybe that's insignificant enough for us to not worry about, but there is a use case difference with SSR which potentially makes this a bigger problem than it has been in the past.

I'm a bit torn between `SSGFallback.None` and `SSGFallback.SSR` as a result. I feel the former is more intuitive to me as a developer, but the latter is more internally consistent. I'm curious to hear what others think? /cc @AndrewKushnir @atscott

Reply:
I don't have a strong preference for the default setting, but I believe SSR is a good default. Next.js also uses a similar "blocking" default, where the page is rendered on the server upon the initial request and then cached for subsequent visits.

If SSR is chosen as a fallback, the same Angular routing rules and guards that apply to CSR are maintained. So, if a product page is accessible via SSR, it will also be accessible via CSR.